[ADD] Added 3d cores to neuralpredictors

Mvystrcilova opened the pull request:
Added Basic3dCore and Factorized3dCore to have cores that support 3d convolutions. In `neuralpredictros/layers/encoders/encoder3d.py` is now Encoder3d as a sample how to put together a 3d core and 2d gaussian readout. Along with the cores there is now 1d laplace in regularizers to smoothen across the temporal dimension and Bias3dLayer and Scale3dLayer.

pollytur commented:
@Mvystrcilova Thanks a lot for your contribution!  
I have some minor comments / questions 

- I would move class Core3d(Core) from [neuralpredictors/layers/cores/conv3d.py](https://github.com/sinzlab/neuralpredictors/pull/196#diff-1d1fc960e3bf20dfa59c5deec837bca2f67b3874609d8cd2d5934c89eb6199c8) to `neuralpredictors/layers/cores/base.py` as its the base class for 3D cores.
- why `Factorized3dCore(nn.Module)` does not inherit from `Core3d`? I would inherit from it and add `self.initialize(cuda=cuda)` line at the end of the `__init__` function to have consistent and fixed initialisations. Or is there any reason for it?

Mvystrcilova commented:
> @Mvystrcilova Thanks a lot for your contribution! I have some minor comments / questions
> 
> * I would move class Core3d(Core) from [neuralpredictors/layers/cores/conv3d.py](https://github.com/sinzlab/neuralpredictors/pull/196#diff-1d1fc960e3bf20dfa59c5deec837bca2f67b3874609d8cd2d5934c89eb6199c8) to `neuralpredictors/layers/cores/base.py` as its the base class for 3D cores.
> * why `Factorized3dCore(nn.Module)` does not inherit from `Core3d`? I would inherit from it and add `self.initialize(cuda=cuda)` line at the end of the `__init__` function to have consistent and fixed initialisations. Or is there any reason for it?

No reason, I just forgot about it. I changed it as you suggested now.

pollytur commented:
Looks fine for me. @cblessing24 could you take a look as well? I do not have the rights to approve the PR

pollytur commented:
@Mvystrcilova and why there is no group sparsity for 3D cores? (Like [here](https://github.com/sinzlab/neuralpredictors/blob/24a954dacb3226ecd60b22573f686fccd0cffce5/neuralpredictors/layers/cores/conv2d.py#L295) for 2D cores) Have you tested it and it did not worked fine or its was never implemented yet? (If it was never implemented yet - its ok, we will just add it a separate PR later if it works nicely)

Mvystrcilova commented:
> @Mvystrcilova and why there is no group sparsity for 3D cores? (Like [here](https://github.com/sinzlab/neuralpredictors/blob/24a954dacb3226ecd60b22573f686fccd0cffce5/neuralpredictors/layers/cores/conv2d.py#L295) for 2D cores) Have you tested it and it did not worked fine or its was never implemented yet? (If it was never implemented yet - its ok, we will just add it a separate PR later if it works nicely)

It was never implemented.

cblessing24 commented on neuralpredictors/layers/cores/conv3d.py:
```diff
@@ -0,0 +1,503 @@
+
+        self.features.add_module("layer0", nn.Sequential(layer))
+
+        for l in range(0, self.layers - 1):
+            layer = OrderedDict()
+            layer[f"conv_{l + 1}"] = nn.Conv3d(
+                self.hidden_channels[l],
+                self.hidden_channels[l + 1],
+                kernel_size=self.hidden_kernel[l],
+                dilation=spatial_dilation,
+                stride=(1, self.stride[l + 1], self.stride[l + 1]),
+                padding=(0, (self.hidden_kernel[l][1]) // 2, (self.hidden_kernel[l][2]) // 2) if self.padding else 0,
+            )
```
This could be simplified using `itertools.pairwise`.

cblessing24 commented on neuralpredictors/layers/cores/conv3d.py:
```diff
@@ -0,0 +1,503 @@
+                padding=(0, (self.hidden_kernel[l][1]) // 2, (self.hidden_kernel[l][2]) // 2) if self.padding else 0,
+            )
+
+            if batch_norm:
+                if independent_bn_bias:
+                    layer["norm"] = nn.BatchNorm3d(hidden_channels, momentum=momentum)
+                else:
+                    layer["norm"] = nn.BatchNorm3d(hidden_channels, momentum=momentum, affine=bias and batch_norm_scale)
+                    if bias:
+                        if not batch_norm_scale:
+                            layer["bias"] = Bias3DLayer(hidden_channels)
+                    elif batch_norm_scale:
+                        layer["scale"] = Scale3DLayer(hidden_channels)
```
Isn't this the same code as above?

cblessing24 commented on neuralpredictors/layers/cores/conv3d.py:
```diff
@@ -0,0 +1,503 @@
+        self.initialize(cuda=cuda)
+
+    def forward(self, x):
+        for l, features in enumerate(self.features):
+            x = features(x)
+        return x
```
Why are you enumerating here instead of just iterating regularly?

cblessing24 commented on neuralpredictors/layers/cores/conv3d.py:
```diff
@@ -0,0 +1,503 @@
+
+    def laplace_temporal(self):
+        laplace = 0
+        for w in range(self.features[0].conv.weight.shape[-2]):
+            for h in range(self.features[0].conv.weight.shape[-1]):
```
You could use `itertools.product` here to combine the two loops.

cblessing24 commented on neuralpredictors/layers/cores/conv3d.py:
```diff
@@ -0,0 +1,503 @@
+            w = math.floor((w - (kernel_size[2] - 1) * spatial_dilation - 1) / self.stride[i] + 1)
+
+        output_shape = self.hidden_channels[layers - 1], t, h, w
+        print("core output shape:", output_shape)
```
Is this left over from debugging? Generally using `print` in library code is not recommended. That is what loggers are for.

cblessing24 commented on neuralpredictors/regularizers.py:
```diff
@@ -291,3 +295,27 @@ def forward(self, x, avg=False):
+        return F.conv1d(x, self.filter, bias=None, padding=self.padding_size)
+
+
+class TimeLaplaceL23d(nn.Module):
```
Is this class name correct?

cblessing24 replied:
I was asking about the name originally because the end looked like it was a typo to me at the time but know I see that that is not the case. So nevermind.

cblessing24 commented on neuralpredictors/regularizers.py:
```diff
@@ -291,3 +295,27 @@ def forward(self, x, avg=False):
+
+class Laplace1d(nn.Module):
+    def __init__(self, padding):
+        super(Laplace1d, self).__init__()
```
Same here.

Mvystrcilova replied:
Yes, why wouldn't it be the correct class name? Or do you mean the arguments in super()?

cblessing24 replied:
Yes, I was talking about the arguments to `super`.

cblessing24 commented on neuralpredictors/layers/encoders/encoder3d.py:
```diff
@@ -0,0 +1,26 @@
+            self.nonlinearity = core.nonlinearities[readout_nonlinearity](elu_xshift, elu_yshift)
+        else:
+            self.nonlinearity = core.nonlinearities[readout_nonlinearity]()
+        self.visualization_dir = None
```
What is the purpose of this attribute?

cblessing24 commented on test/layers/cores/test_conv3d.py:
I think it is great that you were thinking about adding tests to this code but an empty file does not seem useful. :P

cblessing24 commented on neuralpredictors/layers/cores/conv3d.py:
Some of the comments I left on the basic core also apply to the factorized core. I did not look into that one in detail because it seems to contain quite a bit of duplicated code from the basic core. It would be great if that could somehow be reduced...

Mvystrcilova replied:
I agree, but if possible I would do some code merging in a potential future PR and not in this one.
I tried to remove the stuff you marked from both cores when present.

Mvystrcilova commented on neuralpredictors/layers/cores/conv3d.py:
```diff
@@ -0,0 +1,503 @@
+            if input_regularizer == "GaussianLaplaceL2"
+            else dict(padding=laplace_padding)
+        )
+        self._input_weight_regularizer = regularizers.__dict__[input_regularizer](**regularizer_config)
```
This is consistent with regularizer specification in the conv2d cores. That's why it's this way.
